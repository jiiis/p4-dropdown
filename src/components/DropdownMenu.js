import React, { useEffect, useState } from 'react'
import styled from 'styled-components'

const StyledDropdownMenu = styled.div`
  margin: 12px 0 0;
  border: 1px solid #ccc;
  
  & > form {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    padding: 6px;
    height: 42px;
    border-bottom: 1px solid #ccc;
  }
  
  & > form > input {
    flex-grow: 1;
    margin: 0 0 0 6px;
    height: 30px;
    line-height: 30px;
    border: 1px solid #ccc;
  }
  
  li {
    height: 30px;
    line-height: 30px;
    padding: 0 6px;
    cursor: pointer;

    &:hover {
      background: #ccc;
    }
  }
`

const StyledEmptyState = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 72px;
`

const DropdownMenu = ({ availableOptions = [], selectOption }) => {
  // States
  const [searchTerm, setSearchTerm] = useState('')
  const [visibleOptions, setVisibleOptions] = useState(availableOptions)

  // Lifecycle events
  useEffect(() => {
    if (/^\s*$/.test(searchTerm)) {
      setVisibleOptions(availableOptions)

      return
    }

    setVisibleOptions(availableOptions.filter(option => option.label.toLowerCase().includes(searchTerm.toLowerCase())))
  }, [searchTerm, availableOptions])

  // Event handlers
  const handleSearchTextFieldChange = (event) => {
    setSearchTerm(event.target.value)
  }

  return (
    <StyledDropdownMenu>
      <form>
        <label htmlFor="option-search-text-field">Search options</label>
        <input
          id="option-search-text-field"
          type="text"
          name="option-search-text-field"
          value={searchTerm}
          onChange={handleSearchTextFieldChange}
          data-testid="option-search-text-field"
        />
      </form>
      <ul data-testid="visible-option-list">
        {visibleOptions.length === 0 ? <StyledEmptyState>No options available.</StyledEmptyState> : visibleOptions.map(option => (
          <li
            key={option.value}
            onClick={() => selectOption(option.value)}
          >{option.label}</li>
        ))}
      </ul>
    </StyledDropdownMenu>
  )
}

export default DropdownMenu
