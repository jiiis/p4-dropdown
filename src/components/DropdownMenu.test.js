import ReactDom from 'react-dom'
import renderer from 'react-test-renderer'
import { cleanup, render, fireEvent } from '@testing-library/react'

import DropdownMenu from './DropdownMenu'

const availableOptions = [
  {
    label: 'China',
    value: 'china'
  },
  {
    label: 'United States',
    value: 'united-states'
  },
  {
    label: 'Japan',
    value: 'japan'
  }
]

let selectOptionSpy, dropdownMenuComponent

beforeEach(() => {
  selectOptionSpy = jest.fn()
  dropdownMenuComponent = <DropdownMenu
    availableOptions={availableOptions}
    selectOption={selectOptionSpy}
  />
})

afterEach(cleanup)

test('renders in the DOM', () => {
  const div = document.createElement('div')

  ReactDom.render(dropdownMenuComponent, div)
})

test('matches snapshot', () => {
  const tree = renderer.create(dropdownMenuComponent).toJSON()

  expect(tree).toMatchSnapshot()
})

test('filters by search term', () => {
  const { getByTestId } = render(dropdownMenuComponent)

  fireEvent.change(getByTestId('option-search-text-field'), { target: { value: 'hin' } })

  expect(getByTestId('visible-option-list')).toHaveTextContent('China')
  expect(getByTestId('visible-option-list')).not.toHaveTextContent('United States')
  expect(getByTestId('visible-option-list')).not.toHaveTextContent('Japan')
})

test('deselects option', () => {
  const { getByText } = render(dropdownMenuComponent)

  fireEvent.click(getByText('Japan'))

  expect(selectOptionSpy).toHaveBeenCalledWith('japan')
})
