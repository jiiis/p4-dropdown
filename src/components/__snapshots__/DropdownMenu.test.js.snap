// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="sc-bdfBwQ iPcBUy"
>
  <form>
    <label
      htmlFor="option-search-text-field"
    >
      Search options
    </label>
    <input
      data-testid="option-search-text-field"
      id="option-search-text-field"
      name="option-search-text-field"
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
  <ul
    data-testid="visible-option-list"
  >
    <li
      onClick={[Function]}
    >
      China
    </li>
    <li
      onClick={[Function]}
    >
      United States
    </li>
    <li
      onClick={[Function]}
    >
      Japan
    </li>
  </ul>
</div>
`;
